### DISCUSSION TITLE
It is possible to view Javadoc hovers with eclipse.jdt.ls?

### DISCUSSION THREAD

**Discussion by francesco-cadei** [joaotavora/eglot#1364] [Q&A] [answered]
I have installed eglot and setted up with [eclipse.jdt.ls ](https://github.com/eclipse-jdtls/eclipse.jdt.ls).
All works fine but it would be amazing if there was the possibility to view Javadoc hovers like promised by eclipse.jdt.ls.

**Comment by yveszoundi** [comment]
@francesco-cadei , I think that the notion of "hover" in eglot means at the current "cursor location" (i.e., not mouse hover).
I don't know if that was also your current understanding or expectation (_eldoc_).

I've noticed that few "weird" JDT.LS things in respect to javadoc information:
- For basic JDK types, you can usually get javadoc information on methods (`getName`, but not non classes (`java.io.File`)
- For external libraries, it seems that the javadoc information is looked up properly with the cursor either on a method, class or package/module
- There might be other discrepancies, in comparison to what you see in a full blown IDE:  I usually write code without IDE or LSP myself

In the screenshot below, the mouse cursor is on `assertThat` (JUnit method), and the javadoc is displayed inside the `eldoc buffer` (`M-x eldoc-doc-buffer`).
![eglot-hover-java](https://github.com/joaotavora/eglot/assets/5206067/e57cdb70-3a2a-4dcb-abf2-5580a1e49861)

**Reply by francesco-cadei** [reply]
Thanks for the answer.
It is possible to have a little description of the methods like oracle [https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/PrintStream.html#println()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/PrintStream.html#println())?

**Reply by yveszoundi** [reply]
Yes, per screenshot below with the specific example in your follow-up question.
- Does it not work when you try it, with the mouse cursor on `println` and the `eldoc` buffer open? (`M-x eldoc-doc-buffer`)?
- You might also need the package `markdown-mode` installed too, for rendering the javadocs (I can't recall).


![eglot-java-ps-println](https://github.com/joaotavora/eglot/assets/5206067/7e0f9188-a504-491e-9b48-5cfffaa74701)

**Reply by francesco-cadei** [reply]
I have the markdown-mode. But for the `println()` I only have this:

![image](https://github.com/joaotavora/eglot/assets/25520099/08c1035f-0c7c-4473-bc38-88e5130aaa99)

my settings of eglot:
```
(use-package eglot
  :config
  (add-to-list 'eglot-server-programs
	       '(java-mode . ("~/jdt-language-server-1.33.0-202402151717/bin/jdtls")))
  :hook
  (java-mode . eglot-ensure) ;; https://download.eclipse.org/jdtls/milestones/?d
  (js-mode . eglot-ensure) ;; npm install -g typescript-language-server typescript
  )
```

**Reply by yveszoundi** [reply, marked as answer]
I think that you also need to install JDK sources depending on your environment (`openjdk-21-source` on Ubuntu, essentially having `src.zip` in `$JAVA_HOME/jre/lib` accordingly to your JDK version).

**Reply by francesco-cadei** [reply]
I am in debian and I was using `default-jdk`.
Now solved with  `openjdk-17-jdk` and `openjdk-17-source`.
